Software/Rig/Rig.ino

```diff
@@ -0,0 +1,339 @@
+ *
+ *  - Choose to make water available
+ *  - Deliver a positive or negative auditory signal
+ *  - Deliver air puffs or not (positive auditory signal)
+ *
+ * The current implementation makes some assumptions:
+ *
+ *  - Each session is triggered with a button press
+ *  - Air puffs are delivered at a fixed point in the trial
+ *  - There are two auditory signals: positive and negative
+ *    - Positive is a constant 5kHz tone
+ *    - Negative is a pulsed 2kHz tone
+ *  - Air puffs are only delivered after a positive auditory signal
+ *
+ * It can be improved by:
+ *
+ *  - Adding an `init.h` with the values we define at start
+ *  - (So they are not re-used in flush functions)
+ *  - Per-session randomization for
+ *    - Order of set of cues {CS+, CS+, CS+, CS-, CS-, CS-}
+ *    - Inter-trial interval length (e.g. 1-5min)
+ *  - Consistency between "end" and "start" semantics
+ *  - Check that differences in "global time" and "trial time" aren't
+ *    going to cause any issues (some code "thinks" in global, other code
+ *    "thinks" in trial, depending on requirements)
+ *    - For example, refactor air puffs to use just one
+ *
+ */
+#include "rig.h"
+#include "trial.h"
+
+bool sessionHasStarted = false;
+long sessionStartTime = __LONG_MAX__;
+bool sessionHasEnded = false;
+long sessionEndTime = 0;
+
+int currentTrial = 0;
+bool trialHasStarted = false;
+long trialStartTime = __LONG_MAX__;
+bool trialHasEnded = false;
+long trialEndTime = 0;
+
+long lastLickTime = 0;
+bool isLicking = false;
+long waterStartTime = __LONG_MAX__;
+
+bool isPuffing = false;
+long puffStartTime = __LONG_MAX__;
+long puffStopTime = 0;
+
+bool positiveSignalPlaying = false;
+long positiveSignalStart = __LONG_MAX__;
+
+bool negativeSignalPlaying = false;
+bool negativeSignalStart = __LONG_MAX__;
+bool negativeSignalStop = 0;
+
+void setup() {
+  Serial.begin(BAUD_RATE);
+  pinMode(PIN_AIR_PUFF, OUTPUT);
+  pinMode(PIN_WATER_SOLENOID, OUTPUT);
+  pinMode(PIN_TONE_NEGATIVE, OUTPUT);
+  pinMode(PIN_TONE_POSITIVE, OUTPUT);
+  pinMode(PIN_BUTTON, INPUT_PULLUP);
+  pinMode(PIN_LICK, INPUT);
+}
+
+void loop() {
+  if ((digitalRead(PIN_BUTTON) == LOW) && (!sessionHasStarted) && (!sessionHasEnded)) {
+    print("Session has started");
+    sessionHasStarted = true;
+    sessionStartTime = millis();
+  }
+
+  // The session loop. It executes all `N_TRIALS` trials with the
+  // `INTER_TRIAL_WAIT_INTERVAL` wait period between trials.
+  if (sessionHasStarted && !sessionHasEnded) {
+    if (!trialHasStarted) {
+      trialStart();
+    } else {
+      if (!trialHasEnded) {
+        // Core rig loop.
+        // Executes continuously during each trial
+        checkForTrialEnd();
+        checkLick();
+        if (WATER_REWARD_AVAILABLE) {
+          checkWater();
+        }
+        if (DELIVER_AIR_PUFFS) {
+          checkAir();
+          checkPositiveSignal();
+        } else {
+          checkNegativeSignal();
+        }
```

Comment:
These checks (water, air, positive, negative) I was not able to test with the board. But the session (button press), trial looping, and lick timeout I was able to test